libdrgn/kdump.c

```diff
@@ -126,9 +126,72 @@ struct drgn_error *drgn_program_set_kdump(struct drgn_program *prog)
 
 	prog->flags |= DRGN_PROGRAM_IS_LINUX_KERNEL;
 	drgn_program_set_platform(prog, &platform);
+	prog->kdump_ctx = ctx;
 	return NULL;
 
 err:
 	kdump_free(ctx);
 	return err;
 }
+
+struct drgn_error *drgn_program_prstatus_kdump_iter(struct drgn_program *prog,
+						     drgn_prstatus_callback_fn cb_fn)
+{
+	kdump_num_t ncpus;
+	kdump_status ks;
+
+	ks = kdump_get_number_attr(prog->kdump_ctx, "cpu.number", &ncpus);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+					 "kdump_get_number_attr(cpu.number): %s",
+					 kdump_get_err(prog->kdump_ctx));
+	}
+
+	for (int i = 0; i < ncpus; i++) {
+		char *attr_name = NULL;
+		kdump_attr_ref_t prstatus_ref;
+		kdump_attr_t prstatus_attr;
+		unsigned char *prstatus_data;
+		size_t prstatus_size;
+		struct drgn_prstatus_map_entry entry;
+		struct drgn_error *err = NULL;
+		int ret;
+
+		ret = asprintf(&attr_name, "cpu.%d.PRSTATUS", i);
+		if (ret == -1)
+			return &drgn_enomem;
+
+		ks = kdump_attr_ref(prog->kdump_ctx, attr_name, &prstatus_ref);
+		if (ks != KDUMP_OK) {
+			err = drgn_error_format(DRGN_ERROR_OTHER,
+						"kdump_attr_ref(%s): %s",
+						attr_name,
+						kdump_get_err(prog->kdump_ctx));
+			goto kdump_ref_err;
+		}
+
+		ks = kdump_attr_ref_get(prog->kdump_ctx,
+					&prstatus_ref, &prstatus_attr);
+		if (ks != KDUMP_OK) {
+			err = drgn_error_format(DRGN_ERROR_OTHER,
+						"kdump_attr_ref_get(%s): %s",
+						attr_name,
+						kdump_get_err(prog->kdump_ctx));
+			goto kdump_ref_get_err;
+		}
+
+		prstatus_data = kdump_blob_pin(prstatus_attr.val.blob);
+		prstatus_size = kdump_blob_size(prstatus_attr.val.blob);
+
+		err = cb_fn(prog, prstatus_data, prstatus_size);
+
+		kdump_blob_unpin(prstatus_attr.val.blob);
+kdump_ref_get_err:
+		kdump_attr_unref(prog->kdump_ctx, &prstatus_ref);
```

Comment:
Did you confirm with Petr that the unpin/unref is safe?

Reply:
It is not unfortunately. I changed the PR so we don't unpin the blob until the very end. That said, I think it may be better to malloc() separate buffers to copy the PRSTATUS notes from the kdump - the main reason for me thinking that is the fact that we unref the prstatus_ref here but we unpin its blobs later in teardown.

Reply:
As discussed privately, we leave these pins until they are cleaned by `kdump_free()`.